c/shared/source/uforead/uforead.c

```diff
@@ -1912,6 +1801,241 @@ static int parseFontInfo(ufoCtx h) {
+        char* newString = memNew(h, strlen(cpy) + strlen(keyValue) + 2);
+        /* set the 0xC2 to NULL to terminate the left side of the string */
+        *copySymbol = '\0';
+        /* use copySymbol + 2 to skip the NULL and the 0xA9
+           to get the right side of the string */
+        sprintf(newString, "%s%s%s", keyValue, "Copyright", copySymbol + 2);
+        top->Notice.ptr = newString;
+    }
+}
+
+static void setFontInfoVersionMajor(ufoCtx h, char* keyValue) {
+    abfTopDict* top = &h->top;
+
+    if (top->version.ptr == NULL)
+        top->version.ptr = keyValue;
+    else {
+        char* newString = memNew(h, strlen(top->version.ptr) + strlen(keyValue) + 2);
+        sprintf(newString, "%s.%s", keyValue, top->version.ptr);
+        memFree(h, top->version.ptr);
+        top->version.ptr = newString;
+    }
+}
+
+static void setFontIndoVersionMinor(ufoCtx h, char* keyValue) {
+    abfTopDict* top = &h->top;
+
+    if (top->version.ptr == NULL)
+        top->version.ptr = keyValue;
+    else {
+        char* newString = memNew(h, strlen(top->version.ptr) + strlen(keyValue) + 2);
+        sprintf(newString, "%s.%s", top->version.ptr, keyValue);
+        memFree(h, top->version.ptr);
+        top->version.ptr = newString;
+    }
+}
+
+static void setFontInfoUnitsPerEm(ufoCtx h, char* keyValue) {
+    double ppem = strtod(keyValue, NULL);
+    abfTopDict* top = &h->top;
+    abfFontDict* fd = h->top.FDArray.array + h->parseState.GLIFInfo.currentiFD;
+
+    top->sup.UnitsPerEm = (int)ppem;
+    fd->FontMatrix.cnt = 6;
+    fd->FontMatrix.array[0] = (float)(1.0 / ppem);
+    fd->FontMatrix.array[1] = 0;
+    fd->FontMatrix.array[2] = 0;
+    fd->FontMatrix.array[3] = (float)(1.0 / ppem);
+    fd->FontMatrix.array[4] = 0;
+    fd->FontMatrix.array[5] = 0;
+}
+
+static void setFontInfoStdHW(ufoCtx h, char* keyName, char* keyValue) {
+    abfFontDict* fd = h->top.FDArray.array + h->parseState.GLIFInfo.currentiFD;
+    abfPrivateDict* pd = &fd->Private;
+
+    if (keyValue != NULL) {
+       pd->StdHW = (float)strtod(keyValue, NULL);
+    } else {
+       pd->StdHW = (float)strtod(h->valueArray.array[0], NULL);
+       freeValueArray(h);
+    }
+}
+
+static void setFontInfoStdVW(ufoCtx h, char* keyName, char* keyValue) {
+    abfFontDict* fd = h->top.FDArray.array + h->parseState.GLIFInfo.currentiFD;
+    abfPrivateDict* pd = &fd->Private;
+
+    if (keyValue != NULL) {
+        pd->StdVW = (float)strtod(keyValue, NULL);
+    } else {
+        pd->StdVW = (float)strtod(h->valueArray.array[0], NULL);
+        freeValueArray(h);
+    }
+}
+
+static bool setFontInfoFD(ufoCtx h, char* keyName, char* keyValue) {
+    abfTopDict* top = &h->top;
+    abfFontDict* fd = h->top.FDArray.array + h->parseState.GLIFInfo.currentiFD;
+    abfFontMatrix* fontMatrix;
+
+    if (strEqual(keyName, "FontName")) {
+       fd->FontName.ptr = keyValue;
+    } else if (strEqual(keyName, "unitsPerEm")) {
+       setFontInfoUnitsPerEm(h, keyValue);
```

Comment:
What guarantees that the `unitsPerEm` key will be before the `FontMatrix` key in a UFO dictionary? If `unitsPerEm` happens to come later, won't we overwrite the matrix with the identity matrix? 

Could we just initialize the matrix to the identity at some earlier point and not do that in `setFontInfoUnitsPerEm()`? That seems like the cleanest work-around if it's a possibility.

Reply:
@skef just fixed this so that before setting values into fontMatrix when parsing UnitsPerEm key, it checks to see if values have previously been set. It does not set values if previously set (by the FontMatrix key, for example) so that these values are not overwritten.